plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -310,6 +341,25 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session) || !jdbcClient.supportsTopNPushDown()) {
+            return Optional.empty();
+        }
+
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+        List<SortItem> sortOrder = sortItems.stream()
+                .map(sortItem -> new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder()))
+                .collect(toImmutableList());
+        return Optional.of(new TopNApplicationResult<>(flushTopNAttributesAsQuery(session, handle, sortOrder, topNCount), jdbcClient.isLimitGuaranteed(session)));
```

Comment:
`isLimitGuaranteed` is for LIMIT, and may or may not be suitable for TopN pushdown.
we should have a separate method for that.